Тетрадка предназначена для студентов образвательного проекта Booster MDS использование вне проекта запрещено.

##### Задания не несут смысловой нагрузки для реальной жизни, они сделаны лишь для того, чтоб научить использовать Pandas

#### Каждое задание оценивает в 1 балл, для успешной сдачи задания нужно набрать 12 баллов

This database contains 76 attributes, but all published experiments refer to using a subset of 14 of them. In particular, the Cleveland database is the only one that has been used by ML researchers to this date. The "goal" field refers to the presence of heart disease in the patient. It is integer valued from 0 (no presence) to 1.

https://www.kaggle.com/ronitf/heart-disease-uci

___

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('d:\heart_pandas.csv')

In [3]:
data.head(5)

age     sex   chest_pain_type  resting_blood_pressure  cholesterol  \
0   63    male  non-anginal pain                     145          233   
1   37    male   atypical angina                     130          250   
2   41  female    typical angina                     130          204   
3   56    male    typical angina                     120          236   
4   57  female    typical angina                     120          354   

     fasting_blood_sugar               rest_ecg  max_heart_rate_achieved  \
0  greater than 120mg/ml                 normal                      150   
1    lower than 120mg/ml  ST-T wave abnormality                      187   
2    lower than 120mg/ml                 normal                      172   
3    lower than 120mg/ml  ST-T wave abnormality                      178   
4    lower than 120mg/ml  ST-T wave abnormality                      163   

  exercise_induced_angina  st_depression   st_slope  num_major_vessels  \
0                      no            2.3  upsloping                  0   
1                      no            3.5  upsloping                  0   
2                      no            1.4       flat                  0   
3                      no            0.8       flat                  0   
4                     yes            0.6       flat                  0   

    thalassemia  target  
0        normal       1  
1  fixed defect       1  
2  fixed defect       1  
3  fixed defect       1  
4  fixed defect       1

#### Описание признаков

**age**: The person's age in years

**sex**: The person's sex (1 = male, 0 = female)

**cp**: The chest pain experienced (Value 1: typical angina, Value 2: atypical angina, Value 3: non-anginal pain, Value 4: asymptomatic)

**trestbps**: The person's resting blood pressure (mm Hg on admission to the hospital)

**chol**: The person's cholesterol measurement in mg/dl

**fbs**: The person's fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)

**restecg**: Resting electrocardiographic measurement (0 = normal, 1 = having ST-T wave abnormality, 2 = showing probable or definite left ventricular hypertrophy by Estes' criteria)

**thalach**: The person's maximum heart rate achieved

**exang**: Exercise induced angina (1 = yes; 0 = no)

**oldpeak**: ST depression induced by exercise relative to rest ('ST' relates to positions on the ECG plot. See more here)

**slope**: the slope of the peak exercise ST segment (Value 1: upsloping, Value 2: flat, Value 3: downsloping)

**ca**: The number of major vessels (0-3)

**thal**: A blood disorder called thalassemia (3 = normal; 6 = fixed defect; 7 = reversable defect)

**target**: Heart disease (0 = no, 1 = yes)

> 1. age 
> 2. sex 
> 3. chest pain type (4 values) 
> 4. resting blood pressure 
> 5. serum cholestoral in mg/dl 
> 6. fasting blood sugar > 120 mg/dl
> 7. resting electrocardiographic results (values 0,1,2)
> 8. maximum heart rate achieved 
> 9. exercise induced angina 
> 10. oldpeak = ST depression induced by exercise relative to rest 
> 11. the slope of the peak exercise ST segment 
> 12. number of major vessels (0-3) colored by flourosopy 
> 13. thal: 3 = normal; 6 = fixed defect; 7 = reversable defect

___

**Сколько мужчин в датасете? Сколько женщен? (sex) **

In [4]:
res = data['sex'].value_counts()
print('мужчин -',res[0], '; женщин -', res[1])

мужчин - 207 ; женщин - 96


Ответ: мужчин - 207, женщин - 96

 **Какой процент мужчин в датасете? (решите в одну строчку, не используя предыдущий результат. Не забудте знак процента) **

In [5]:
print(data['sex'].value_counts('male')['male']*100,'%')

68.3168316832 %


или

In [6]:
'{:.2%}'.format(data['sex'].value_counts()[0] / data.shape[0])

'68.32%'

Ответ: мужчин в датасете 68.32%

**Сколько мужчин имеют заболевание сердца? Сколько женщин имеют заболевание сердца? **

In [7]:
men_with_desease = data[(data['sex'] == 'male') & (data['target'] == 1)].shape[0]
women_with_desease = data[(data['sex'] == 'female') & (data['target'] == 1)].shape[0]
print('мужчин с заболеванием сердца: ', men_with_desease, ', женщин с заболеванием сердца: ', women_with_desease)

мужчин с заболеванием сердца:  93 , женщин с заболеванием сердца:  72


Ответ: мужчин - 93, женщин - 72

**Какую долю, от общего числа пациентов, занимают мужчины не имеющие болезнь сердца?**

In [8]:
data[(data['sex'] == 'male') & (data['target'] == 0)].shape[0] / data.shape[0]

0.37623762376237624

Ответ: 0.38 от всей выборки

**Сколько лет самому молодому пациенту, страдающему болезнью сердца?**

In [9]:
data[data['target'] == 1]['age'].min()

29

Ответ: 29 лет

**Сколько лет самому возастному пациенту, у которого нет проблем с сердцем?**

In [10]:
data[data['target'] == 0]['age'].max()

77

Ответ: 77 лет

**Сколько лет самой молодой женщине, которая страдает болезнью сердца?**

In [11]:
print('возраст самой молодй женщины: ', data[(data['target'] == 1) & (data['sex'] == 'female')]['age'].min())

возраст самой молодй женщины:  34


Ответ: 34 года

**Какой средний возраст женщин?**

In [12]:
print('средний возраст женщин: ', data[data['sex'] == 'female']['age'].mean())

средний возраст женщин:  55.6770833333


Ответ: средний возраст всех женщин из датасета - 56 лет

**Каковы средние значения и среднеквадратичные отклонения возраста тех, кто страдают болезнью сердца?**

In [13]:
print('среднее значение возраста: ', data[data['target'] == 1]['age'].mean(), 
      '\nсреднеквадратичное отклонение возраста: ', data[data['target'] == 1]['age'].std())

среднее значение возраста:  52.496969697 
среднеквадратичное отклонение возраста:  9.55065075195


Ответ: среднее значение возраста:  52.5 года, среднеквадратичное отклонение возраста:  9.55 лет

** Правда ли, что люди не болеющие болезнью сердца имеют уровня холестерина меньше среднего? (chol) **

In [14]:
avg_chol = data['cholesterol'].mean()
avg_chol_for_no_heart_deseases = data[data['target'] == 0]['cholesterol'].mean()
if(avg_chol_for_no_heart_deseases < avg_chol): 
    print("Yes")
else:
    print("No")

No


Ответ: нет, не правда

**Выведите статистику rest_ecg для все числовых признаков, его максимально и среднее значение (используйте groupby(), решите в одну строчку)**

In [16]:
data.groupby(by='rest_ecg')['age', 'resting_blood_pressure', 'cholesterol', \
                            'max_heart_rate_achieved', 'st_depression', 'num_major_vessels'].describe().reset_index()

rest_ecg    age                                     \
                                 count       mean        std   min    25%   
0         ST-T wave abnormality  152.0  52.914474   9.229196  34.0  44.75   
1  left ventricular hypertrophy    4.0  61.000000  10.099505  55.0  55.00   
2                        normal  147.0  55.687075   8.675902  29.0  50.50   

                    resting_blood_pressure ...  st_depression       \
    50%   75%   max                  count ...            75%  max   
0  53.5  60.0  71.0                  152.0 ...           1.40  5.6   
1  56.5  62.5  76.0                    4.0 ...           3.65  4.4   
2  57.0  62.0  77.0                  147.0 ...           1.95  6.2   

  num_major_vessels                                               
              count      mean       std  min  25%  50%  75%  max  
0             152.0  0.638158  1.006925  0.0  0.0  0.0  1.0  4.0  
1               4.0  1.000000  1.414214  0.0  0.0  0.5  1.5  3.0  
2             147.0  0.816327  1.027168  0.0  0.0  0.0  1.0  4.0  

[3 rows x 49 columns]

In [16]:
## Ваш ответ:

**Посчитайте у кого уровень депрессии при физический нагрузке выше (в среднем), среди мужчин страдающих болезнью сердца или среди женщин не страдающих болезнью сердца (st_depression)**

In [17]:
male_lvl_of_depression = data[(data['sex'] == 'male') & (data['target'] == 1)]['st_depression'].mean()
female_lvl_of_depression = data[(data['sex'] == 'female') & (data['target'] == 0)]['st_depression'].mean()
if(male_lvl_of_depression > female_lvl_of_depression):
    print('Уровень депрессии выше среди мужчин с болезнями сердца')
else:
    print('Уровень депрессии выше среди женщин не страдающих болезнями сердца')

Уровень депрессии выше среди женщин не страдающих болезнями сердца


Ответ: Уровень депрессии выше среди женщин не страдающих болезнями сердца

**Посчтитайте максимальный и минимальный уровень холестерина для каждого типа chest_pain_type, rest_ecg, thalassemia. Пишите код оптимально, можно использовать циклы**

In [18]:
my_list = ['chest_pain_type', 'rest_ecg', 'thalassemia']
for var in my_list:
    print ('______________________')
    print ('GROUP BY', var)
    print ('')
    print (pd.pivot_table(data, index = [var], values = ['cholesterol'], aggfunc = [min, max]))

______________________
GROUP BY chest_pain_type

                         min         max
                 cholesterol cholesterol
chest_pain_type                         
atypical angina          126         564
non-anginal pain         182         298
typical angina           131         409
______________________
GROUP BY rest_ecg

                                     min         max
                             cholesterol cholesterol
rest_ecg                                            
ST-T wave abnormality                126         354
left ventricular hypertrophy         197         327
normal                               149         564
______________________
GROUP BY thalassemia

                          min         max
                  cholesterol cholesterol
thalassemia                              
fixed defect              141         417
normal                    169         318
reversable defect         126         564


или если нужно вывести показатели для комбинаций значений этих признаков:

In [26]:
pd.pivot_table(data, index = ['chest_pain_type', 'rest_ecg', 'thalassemia'], \
               values = ['cholesterol'], aggfunc= [min, max])

min  \
                                                                cholesterol   
chest_pain_type  rest_ecg                     thalassemia                     
atypical angina  ST-T wave abnormality        fixed defect              141   
                                              normal                    218   
                                              reversable defect         126   
                 left ventricular hypertrophy fixed defect              197   
                 normal                       fixed defect              149   
                                              normal                    216   
                                              reversable defect         212   
non-anginal pain ST-T wave abnormality        fixed defect              204   
                                              reversable defect         199   
                 normal                       fixed defect              182   
                                              normal                    186   
                                              reversable defect         193   
typical angina   ST-T wave abnormality        fixed defect              149   
                                              normal                    169   
                                              reversable defect         131   
                 left ventricular hypertrophy fixed defect              327   
                                              normal                    318   
                                              reversable defect         205   
                 normal                       fixed defect              166   
                                              normal                    176   
                                              reversable defect         164   

                                                                        max  
                                                                cholesterol  
chest_pain_type  rest_ecg                     thalassemia                    
atypical angina  ST-T wave abnormality        fixed defect              340  
                                              normal                    218  
                                              reversable defect         313  
                 left ventricular hypertrophy fixed defect              197  
                 normal                       fixed defect              417  
                                              normal                    256  
                                              reversable defect         564  
non-anginal pain ST-T wave abnormality        fixed defect              240  
                                              reversable defect         298  
                 normal                       fixed defect              283  
                                              normal                    233  
                                              reversable defect         288  
typical angina   ST-T wave abnormality        fixed defect              354  
                                              normal                    315  
                                              reversable defect         353  
                 left ventricular hypertrophy fixed defect              327  
                                              normal                    318  
                                              reversable defect         205  
                 normal                       fixed defect              394  
                                              normal                    276  
                                              reversable defect         409

In [ ]:
## Ваш ответ:

**Сколько значений может принимать каждый из категориальных признаков?**

In [27]:
#my_list = ['sex', 'chest_pain_type', 'fasting_blood_sugar', 'rest_ecg', 'exercise_induced_angina', 'st_slope', 'thalassemia','target']
my_list = [c for c in data.columns if data[c].dtype.name == 'object']
my_list.append('target')
for var in my_list:
    print ('___________________')
    print ('VAR CAT', var)
    print (data[var].unique())
    print (data[var].unique().shape[0])

___________________
VAR CAT sex
['male' 'female']
2
___________________
VAR CAT chest_pain_type
['non-anginal pain' 'atypical angina' 'typical angina']
3
___________________
VAR CAT fasting_blood_sugar
['greater than 120mg/ml' 'lower than 120mg/ml']
2
___________________
VAR CAT rest_ecg
['normal' 'ST-T wave abnormality' 'left ventricular hypertrophy']
3
___________________
VAR CAT exercise_induced_angina
['no' 'yes']
2
___________________
VAR CAT st_slope
['upsloping' 'flat']
2
___________________
VAR CAT thalassemia
['normal' 'fixed defect' 'reversable defect']
3
___________________
VAR CAT target
[1 0]
2


In [28]:
## Ваш ответ:

**У какого категориального признака наблюдается самый сильный дизбаланс классов?**

In [29]:
import math
sum_of_records = data.shape[0]
diction = {}
for var in my_list:
    var_list = []
    unique = data[var].unique()
    for j in unique:
        var_list.append(data[data[var] == j].shape[0])
    var_list.sort()
    min_percent = var_list[0] / sum_of_records
    max_percent = var_list[len(var_list)-1] / sum_of_records
    diction[var] = (max_percent - min_percent)*100
    
max_disbalance = 0
cat_with_max_disbalance = ''

for i in diction.keys():
    if diction[i] > max_disbalance:
        max_disbalance = diction[i]
        cat_with_max_disbalance = i
print(cat_with_max_disbalance)

fasting_blood_sugar


Самый сильный дисбаланс классов наблюдается для признака "fasting_blood_sugar"

---

@Booster
Мамкин Data Scientist

tg @mommyscience